deltacat/compute/compactor_v2/steps/merge.py

```diff
@@ -1,18 +1,439 @@
+        key=lambda df: (df.stream_position, df.file_index),
+        reverse=False,  # ascending
+    )
+    is_delete = False
+    for df_envelope in df_envelopes:
+        assert (
+            df_envelope.delta_type != DeltaType.APPEND
+        ), "APPEND type deltas are not supported. Kindly use UPSERT or DELETE"
+        if df_envelope.delta_type == DeltaType.DELETE:
+            is_delete = True
+
+    for df_envelope in df_envelopes:
+        table = df_envelope.table
+        if is_delete:
+            table = _append_delta_type_column(
+                table, np.bool_(sc.delta_type_to_field(df_envelope.delta_type))
+            )
+
+        hb_tables.append(table)
+
+    result = pa.concat_tables(hb_tables)
+
+    return result
+
+
+def _merge_tables(
+    table: pa.Table,
+    primary_keys: List[str],
+    can_drop_duplicates: bool,
+    compacted_table: Optional[pa.Table] = None,
+) -> pa.Table:
+    """
+    Merges the table with compacted table dropping duplicates where necessary.
+
+    This method ensures the appropriate deltas of types DELETE/UPSERT are correctly
+    appended to the table.
+    """
+
+    all_tables = []
+    incremental_idx = 0
+
+    if compacted_table:
+        incremental_idx = 1
+        all_tables.append(compacted_table)
+
+    all_tables.append(table)
+
+    if not primary_keys or not can_drop_duplicates:
+        logger.info(
+            f"Not dropping duplicates for primary keys={primary_keys} "
+            f"and can_drop_duplicates={can_drop_duplicates}"
+        )
+        all_tables[incremental_idx] = _drop_delta_type_rows(
+            all_tables[incremental_idx], DeltaType.DELETE
+        )
+        # we need not drop duplicates
+        return pa.concat_tables(all_tables)
+
+    all_tables = generate_pk_hash_column(all_tables, primary_keys=primary_keys)
+
+    result_table_list = []
+
+    incremental_table = drop_duplicates(
+        all_tables[incremental_idx], on=sc._PK_HASH_STRING_COLUMN_NAME
+    )
+
+    if compacted_table:
+        compacted_table = all_tables[0]
+
+        records_to_keep = pc.invert(
+            pc.is_in(
+                compacted_table[sc._PK_HASH_STRING_COLUMN_NAME],
+                incremental_table[sc._PK_HASH_STRING_COLUMN_NAME],
+            )
+        )
+
+        result_table_list.append(compacted_table.filter(records_to_keep))
+
+    incremental_table = _drop_delta_type_rows(incremental_table, DeltaType.DELETE)
+    result_table_list.append(incremental_table)
+
+    final_table = pa.concat_tables(result_table_list)
+    final_table = final_table.drop([sc._PK_HASH_STRING_COLUMN_NAME])
+
+    return final_table
+
+
+def _download_compacted_table(
+    hb_index: int,
+    rcf: RoundCompletionInfo,
+    read_kwargs_provider: Optional[ReadKwargsProvider] = None,
+    deltacat_storage=unimplemented_deltacat_storage,
+    deltacat_storage_kwargs: Optional[dict] = None,
+) -> pa.Table:
+    tables = []
+    hb_index_to_indices = rcf.hb_index_to_entry_range
```

Comment:
A RCF may not necessarily have`hb_index_to_entry_range` right?

Reply:
If you are downloading compacted table, there has to be rcf. See the code below that checks this.

Reply:
We might also need a check for `_copy_all_manifest_files_from_old_hash_buckets` as well if we want backwards compatibility with old RCFs that might not have this value.

Reply:
What I meant is, we call this function only if hb_index_to_entry_range is present. 

```python
if ( input.round_completion_info
                    and input.round_completion_info.hb_index_to_entry_range
                    and input.round_completion_info.hb_index_to_entry_range.get(
                        str(hb_idx)
                    )
                    is not None
                )
```

Reply:
For downloading compacted tables, that makes sense. Are we sure we're performing the same kind of check when copying manifests though?

Reply:
Got it.. Addressed.